backends/bmv2/common/action.cpp

```diff
@@ -20,18 +20,20 @@ limitations under the License.
 
 namespace BMV2 {
 
+using namespace P4::literals;
```

Comment:
Do we always need to include this? Maybe we should just define it in the cstring class?

Reply:
Tagging @vlstill This is how the custom string literals are normally done. Standard library also defines lots of custom literals, but they are all guarded by additional namespace (see `Notes` in https://en.cppreference.com/w/cpp/string/basic_string/operator%22%22s as an example).

You cannot define it inside class, it should be globally visible operator.

Reply:
Yeah the reason I am bringing it up is because it might be a little tedious to add this `using` declaration.

Reply:
Yes. Though one could do `cstring("foo")`, or even better `cstring::literal("foo")` instead. IIR

Reply:
Sadly there is no namespace that would all of P4C exist in, so we can only choose between a global operator, or hiding it in a namespace like I did. The problem with global is that there is a chance it will clash with some other operator. Especially if we ever want P4C to be usable as a library, we should not pollute global namespace with operators. Also, this means the using is forbidden on the namespace-level of headers (in headers it can only exist in functions).

But I don't think this would be too bad to use overall. Especially as we try to eliminate unnecessary uses of cstring.

Reply:
I think we should define what is a good practice for using `_cs` literal.

- Is it OK to export it to other namespaces (e.g. `IR`, `BVM2`, etc.)? My gut feeling would be no, but that might be too restrictive for our intentions. These are our namespaces so we can probably afford to export it.
- I would say it is **not OK** to export it to global namespace in a header file. Implementation files can do whatever they can.
  - So a header without a namespace would be limited to using it in functions.

Reply:
> Is it OK to export it to other namespaces (e.g. IR, BVM2, etc.)? My gut feeling would be no, but that might be too restrictive for our intentions. These are our namespaces so we can probably afford to export it.

Here we have quite a common problem: we cannot import it to, say, class, we can only do this into enclosing namespace or at individual function level. The latter is too chatty and too prohibitive. So, we'd end with the cases like:

```c++
namespace Foo {
using namespace P4::literals;

class Bar {
  cstring toString() const { return "Bar"_cs; }
};
}
```

Reply:
> Sadly there is no namespace that would all of P4C exist in, so we can only choose between a global operator, or hiding it in a namespace like I did. The problem with global is that there is a chance it will clash with some other operator. Especially if we ever want P4C to be usable as a library, we should not pollute global namespace with operators. Also, this means the using is forbidden on the namespace-level of headers (in headers it can only exist in functions).

I think this raises a good point, we should like qualify P4C with the P4 namespace to avoid clashes. I already ran into this problem before with a library build. This way we can also use _cs without needing to declare using everywhere.